[STEP 2] 블랙잭 리뷰 요청

y2gcoder opened the pull request:
안녕하세요 한솔님!

지난번 코틀린 DSL 피드백 주신 것 반영해봤는데, 한솔님께서 말씀하셨던 의도가 맞는지 모르겠습니다 🤣

추가로 이번 블랙잭 미션도 구현하여 한솔님의 피드백을 받고자 리뷰 요청을 드립니다! 

감사합니다!

양영근 올림

--- 

# 질문

- 작은 엔티티란 무엇일까요? 최대한 책임을 나눠갖게 해야한다는 의미로 이해하고 각 객체 별로 책임을 나눠갖게 하도록 노력해봤습니다!
- 테스트를 쉽게 하기 위해 Deck.shuffle애 seed를 추가했습니다. Random을 바로 넘기는 게 나을까요?

y2gcoder committed:
refactor(kotlin-dsl): 코틀린 DSL 리뷰어님(한솔님) 피드백 반영

- Test에 모든 객체들을 두었던 것을 다른 파일들로 분리함
- @DslMaker 를 통해 외부 리시버를 사용하지 못하게 막기
- @DisplayName 을 통해 테스트에 대한 정보 명시

y2gcoder committed:
docs(blackjack): 블랙잭 구현 기능 목록 작성

y2gcoder committed:
feature(blackjack): 각 Card에 대한 모양(Suit)을 스페이드, 클로버, 하트, 다이아몬드로 제한

- enum class 활용

y2gcoder committed:
feature(blackjack): 각 Card에 대한 숫자(CardNumber)를 A, 2-10, J, Q, K로 제한

- enum class 활용

y2gcoder committed:
feature(blackjack): CardNumber의 10, J, Q, K 는 모두 10으로 계산하도록 baseValue 설정

y2gcoder committed:
docs(blackjack): 구현 기능 목록 변경

카드 뭉치의 구현 기능 목록 수정

y2gcoder committed:
feature(blackjack): Deck 생성 및 draw() 구현

- Deck에서 mutableList의 removeFirst() 를 통해  draw 구현

y2gcoder committed:
feature(blackjack): Deck 기본 생성시 52장의 카드를 가지고 시작하도록 생성

companion object를 사용해서 기본 카드 리스트 생성 및 기본값에 사용

y2gcoder committed:
feature(blackjack): Deck 의 생성자 유효성 검사 추가

기본값 대신 cards 를 입력할 때도 52장을 넘지 않도록 유효성 검사 추가

y2gcoder committed:
feature(blackjack): Deck 에 카드 순서를 섞을 수 있는 shuffle() 추가

내부에서는 MutableList.shuffle() 사용

y2gcoder committed:
feature(blackjack): Player 추가 및 name 유효성 검사 추가

빈 문자열 체크

y2gcoder committed:
feature(blackjack): Player initialCards  유효성 검사 추가

처음 시작할 때 받는 카드의 개수가 2인지 체크

y2gcoder committed:
feature(blackjack): Player 가 카드 한 장을 받는 receive(newCard: Card) 추가

y2gcoder committed:
feature(blackjack): CardNumber 에 biggerValue() 추가

ACE의 경우 biggerValue가 11, 나머지 CardNumber는 biggerValue가 baseValue와 동일하도록 설정

y2gcoder committed:
feature(blackjack): Player 에 sumOfHand() 추가

손에 들고 있는 모든 cardNumber의 baseValue 의 합을 구함. (ACE 고려하여 21에 가까운 수 고르기 X)

y2gcoder committed:
feature(blackjack): 손패를 뜻하는 일급 컬렉션 Hand 추가 및 Hand에 card를 추가하기 위한 add(newCard: Card) 추가

y2gcoder committed:
feature(blackjack): Hand 에 손패가 가진 카드 합을 구할 수 있는 sumOfHand() 추가

- ACE 카드의 숫자 케이스 미고려

y2gcoder committed:
feature(blackjack): Hand 의 sumOfHand() 에 ACE 카드 케이스 추가

ACE 개수에 따른 테스트 케이스 추가
ACE 개수에 따라 계산 로직 추가:
- ACE 개수는 최대 4개까지 가능(모양에 따라)
- ACE는 최대 1개만 11로 가능함. 2개 이상 11로 했을 때는 블랙잭이 될 수 없음

y2gcoder committed:
feature(blackjack): Player 에 Hand 적용

Player의 손패를 List<Card> 에서 Hand라는 일급 컬렉션으로 변경. Hand에서 필요한 메서드 적용

y2gcoder committed:
feature(blackjack): Hand 에 isBust() 추가

Hand 에 스스로 버스트 했는지 알 수 있는 메서드 isBust() 추가

y2gcoder committed:
feature(blackjack): Player 에 isBust() 추가

Player에도 isBust() 추가(내부적으로는 Hand의 isBust 사용)

y2gcoder committed:
feature(blackjack): 블랙잭 UI 로직 추가

- BlackJackGame: 콘솔 메인 함수
- ConsoleInput: 콘솔 입력
- ConsoleOutput: 콘솔 출력
- DrawAnswer: 콘솔 Y/N

catsbi commented on src/main/kotlin/blackjack/CardNumber.kt:
```diff
@@ -0,0 +1,24 @@
+package blackjack
+
+enum class CardNumber(val face: String, val baseValue: Int) {
```
열거타입을 활용해주셨네요 😄 

다만 다음 미션들부터, 각 카드의 역할이나 책임이 늘어나는 만큼 열거타입으로는 제약이 있지 않을까? 생각이 드네요.

catsbi commented on src/main/kotlin/blackjack/Deck.kt:
```diff
@@ -0,0 +1,30 @@
+
+    fun size(): Int = cards.size
+
+    fun shuffle(seed: Long? = null) {
+        val random = seed?.let { Random(it) } ?: Random.Default
+        cards.shuffle(random)
+    }
```
보통 Shuffler 인터페이스등을 통해 random or fixed 구현체를 선택함으로써 테스트 케이스 구현을 지원했는데, seed값을 통해 값 고정을 지원하도록 하셨군요 👍 

다만, 섞는다는 행위 자체가오직 랜덤한 방식밖에 지원하지 않을 것 같다는 생각이 들어 확장성은 좀 떨어지지 않을까 생각이 드네요

catsbi commented on src/main/kotlin/blackjack/Deck.kt:
```diff
@@ -0,0 +1,30 @@
+
+    companion object {
+        private const val MAXIMUM_CARD_COUNT = 52
+        private val CARDS: List<Card> =
+            Suit.entries.flatMap { suit ->
+                CardNumber.entries.map { cardNumber ->
+                    Card(cardNumber, suit)
+                }
+            }
```
보면 dot(.)이 좀 많이 전개되고 있는 것 같아요. 이렇게 엔티티 그래프 탐색이 깊어지는 경우 적절히 책임 위임을 통해 깊이를 줄여볼 수 있습니다.

catsbi commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -0,0 +1,47 @@
+        if (aceCount == 0) {
+            return nonAceSum
+        }
+        return calculateTotalSumWithAceSum(nonAceSum, aceCount)
```
인터페이스와 구현체를 이용한 추상화를 통한다면 행위(method)를 구분하지 않고 처리할 수도 있을 것 같아요 👍

catsbi commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -0,0 +1,47 @@
+        nonAceSum: Int,
+        aceCount: Int,
+    ): Int {
+        val baseAceSum = aceCount * CardNumber.ACE.baseValue
+        val biggerAceSum = (aceCount - 1) * CardNumber.ACE.baseValue + CardNumber.ACE.biggerValue()
+
+        val baseSum = nonAceSum + baseAceSum
+        val biggerSum = nonAceSum + biggerAceSum
+
+        return if (biggerSum <= BLACKJACK_NUMBER) biggerSum else baseSum
```
우선 ACE의 값이 1일경우와 11일 경우 모두 계산을 수행하고 최종적으로 선택 분기를 고르는 방식이네요 🤔 
규칙이 지금 기준으로 고정이라면 괜찮을 것 같은데, 더 복잡해지는 경우일 경우 Hand라는 클래스의 계산 방식은 유연성을 가지기 힘든 방식으로 보입니다. 좀 더 추상화해볼 수 없을까요?

catsbi commented on src/main/kotlin/blackjack/Player.kt:
```diff
@@ -0,0 +1,32 @@
+        require(name.isNotBlank()) {
+            "이름은 빈 문자열일 수 없습니다: name=$name"
+        }
+        require(initialCards.size == INITIAL_HAND_LENGTH) {
+            "첫 패는 ${INITIAL_HAND_LENGTH}장이어야 한다: initialCards.size=${initialCards.size}"
+        }
```
예외메세지가 충분히 어떤 예외인지에 대한 내용을 내포하고 있어 디버깅 레벨을 낮춰주는 것 같아요 👍

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -0,0 +1,66 @@
+    println()
+
+    players.forEach { player ->
+        while (true) {
```
반복문 중 while을  권장드리진 않고 그 중에서도 중단 플래그가 없는 while문은 더욱 권장드리지 않습니다. 

우선 while문은 플래그가 되는 분기값이 반복문 컨텍스트 내에서소비되는게 아닌 외부 컨텍스트에서 가지고 있기 때문에 플래그의 생명주기가 다른 반복문에 비교해서 더 깁니다. 
그리고 중단 플래그가 없는 반복문은 언제 스택오버플로 예외가 던져질지 모른다는 위험도 있구요.

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -0,0 +1,66 @@
+        printPlayerNameWithHand(player)
+        println()
+    }
+    println()
```
출력에 대해서 출력 계층으로 책임을 위임하기로 했다면, 이런 출력 문구함수 호출이 애플리케이션 레벨에선 별도로 없어도 될 것 같네요.

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -0,0 +1,66 @@
+                    break
+                }
+            }
+        }
```
현재 코드의 indent가 많이 높은것 같아요. 적절히 함수 분리를 하거나해서 indent를 낮춰볼 수 있지 않을까요?

catsbi reviewed (changes requested):
영근님, 2단계 블랙잭 미션, 생각보다 난이도가 있었을텐데 잘 구현해주셨습니다 👍 

몇 가지 추가 고민해보면 좋을 부분들이 있어 코멘트 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요!

y2gcoder committed:
feature(blackjack): Deck의 셔플 방법을 인터페이스 Shuffler로 분리 및 인자로 받은 cards를 deepCopy 후 프로퍼티로 할당

한솔님 피드백 1. shuffle 방식이 랜덤한 것으로 고정되어 확장성이 좋지 않을 것 같다는 의견 반영
: 인터페이스 + 구현체 사용, 생성자, 메서드 모두 주입 가능한 방식
=> 이렇게 하니 Deck에서 Random에 대한 의존이 사라져 좋네요!

y2gcoder committed:
refactor(blackjack): Deck에서 CARDS 상수 생성 방식 리팩토링

한솔님 피드백 2 반영: 카드 전부를 만들어내는 Factory를 만들어 위임하는 방식과 확장함수를 만들어내는 방식 중 확장함수를 만들어내는 방식을 취했습니다!
지난 수업 때 들었던 바를 토대로상수로 전체 카드 목록을 만들게 된 이유가 해당 카드 목록을 지속적으로 재생성하는게 비효율적이라고 생각했기 때문에 확장함수를 이용한 리팩토링을 고안해봤습니다!

y2gcoder committed:
refactor(blackjack): CardNumber 를 sealed class 와 상속 클래스들로 구성 && Hand 에서 카드 합 구하는 책임을 BlackJackCardSumCalculator 로 분리 && Deck 생성시 카드 목록 생성부 리팩토링

한솔님 피드백 3 반영:
한솔님께서 말씀해주신 피드백을 제대로 이해했는지 모르겠습니다.
도와주십쇼!

y2gcoder committed:
refactor(blackjack): BlackJackGame 에서 while(true) 를 generateSequence{}.toList()로 처리

한솔님 피드백 반영: 찝찝했던 while(true) 문을 개선하려고 찾아보다가 sequence를 이용하면 어떨까 해서 해봤습니다!

y2gcoder committed:
refactor(blackjack): BlackJackGame 에서 println() 삭제

최대한 출력문은 ConsoleOutput에서 처리하도록 변경해봤습니다!

y2gcoder commented:
늦어서 죄송합니다!

제가 더 고민하는 것보단 리뷰어님의 피드백이 필요할 것 같아 다시 피드백 요청 드리게 되었습니다!

감사합니다

양영근 올림

catsbi commented on src/main/kotlin/blackjack/Card.kt:
```diff
@@ -1,3 +1,9 @@
+        return number.baseValue
+    }
+
+    override fun compareTo(other: Card): Int = number.compareTo(other.number)
```
소소하지만 compareTo는 infix 함수이기에 다음과 같이 dot을 생략할수도 있습니다 ㅎㅎ

```suggestion
    override fun compareTo(other: Card): Int = number compareTo other.number
```

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -63,4 +39,46 @@ private fun setupPlayers(
+    player: Player,
+    deck: Deck,
+): Unit? {
+    if (isPlayerBust(player)) return null
```
block `{}` 생략은 코드의 일관성을 깨트리고 가독성을 낮추는 문제가 있어 저는 한 줄짜리 코드라도 블럭 생략은 권장드리지 않습니다.

catsbi commented on src/main/kotlin/blackjack/ui/BlackJackGame.kt:
```diff
@@ -63,4 +39,46 @@ private fun setupPlayers(
+    }
+}
+
+private fun isPlayerBust(player: Player): Boolean {
```
ui지만 현재 서비스 레이어에 있어야할 비즈니스 로직들이 있는 것 같아요. 🤔

y2gcoder replied:
저는 단순히 메서드 추출만 했다고 생각했는데 그렇게 하면 비즈니스 로직이 되어버리는군요...
감사합니다! 정말 많이 배우는 것 같습니다! 
다음 피드백 때 한 번 봐주십쇼!

catsbi commented on src/test/kotlin/blackjack/DefaultBlackJackCardSumCalculatorTest.kt:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    "손패는 가진 카드의 숫자 합을 21에 최대한 가깝게 구할 수 있다(Ace 4개)" {
+        forAll(
```
테스트케이스에 픽스쳐가 너무 길게 펼쳐져있다보니 오히려 뭘 테스트하려는건지 의도를 찾기가 힘들어지는 문제가 있는 것 같아요. 적절히 fixture 파일로 분리해보는건 어떨까요?

catsbi commented on src/main/kotlin/blackjack/CardFactory.kt:
```diff
@@ -0,0 +1,16 @@
+    ): Card = Card(cardNumber, suit)
+
+    fun all(): List<Card> {
+        return Suit.entries.flatMap { suit ->
+            CardNumberFactory.all().map { cardNumber ->
+                Card(cardNumber, suit)
+            }
+        }
+    }
```
행위가 factory 객체로 옮겨졌지만 엔티티 그래프 탐색이 깊다는 부분에서는 동일합니다. 코틀린은 함수형 프로그래밍이 수월하기 때문에 좀 더 책임 위임을 하면 어떨까 라는 의견은 팩토리객체로 새로 만들어서 옮기라는 피드백보다는 행위 자체를 더 적절한 객체에 위임시켜서 엔티티 그래프 탐색의 깊이를 낮추라는 의도였습니다. 

```suggestion
        return Suit.flatMap { suit ->
            CardNumberFactory.mapValue { cardNumber ->
                Card(cardNumber, suit)
            }
        }
    }
```

catsbi reviewed (approved):
영근님, 피드백 반영 잘 해주셨어요. 
잘 구현해주셨는데, 몇몇 부분들을 보면 팩토리 패턴을 사용하셨는데, 굳이 팩토리로 책임이 위임되는 문제가 있는 것 같아요. 피드백 확인하셔서 다음 미션 진행하시면서 다시 고민해보면 좋을 것 같습니다. 

추가적인 피드백 확인해주시고 다음 미션에서 같이 고민해보면 좋을 것 같습니다. 2단계는 머지하도록 하겠습니다.
